mono_repo/lib/src/commands/pub.dart

```diff
@@ -5,116 +5,291 @@
+      Iterable<String> defaultsTo,
+      void Function(List<String> values) callback,
+      bool splitCommas = true,
+      bool hide = false}) {}
+
+  /// [_PubArgParser] can't have options added to it, so we do nothing.
+  @override
+  void addOption(String name,
+      {String abbr,
+      String help,
+      String valueHelp,
+      Iterable<String> allowed,
+      Map<String, String> allowedHelp,
+      String defaultsTo,
+      Function callback,
+      bool allowMultiple = false,
+      bool splitCommas,
+      bool hide = false}) {}
+
+  /// [_PubArgParser] can't have separators added to it, so we do nothing.
+  @override
+  void addSeparator(String text) {}
+
+  @override
+  bool get allowTrailingOptions => _delegate.allowTrailingOptions;
+
+  @override
+  bool get allowsAnything => _delegate.allowsAnything;
+
+  @override
+  Map<String, ArgParser> get commands => _delegate.commands;
+
+  @override
+  Option findByAbbreviation(String abbr) => _delegate.findByAbbreviation(abbr);
+
+  /// [_PubArgParser] can't have options, so if asked about a default we return
+  /// null.
+  @override
+  dynamic getDefault(String option) => null;
+
+  @override
+  String getUsage() => _delegate.usage;
+
+  /// Return the options as a [_PubOptions] to trick [CommandRunner].
+  @override
+  Map<String, Option> get options => _PubOptions(_delegate.options);
+
+  /// Wrap the normal output with a `PubArgResults` to fail quieter than normal.
+  @override
+  ArgResults parse(Iterable<String> args) => _delegate.parse(args);
+
+  @override
+  String get usage => '''Any arguments given are passed verbatim to `pub`.\n
+If a particular package uses Flutter, `flutter` is used rather than `pub`:
+- If the arguments begin with `get` or `upgrade`, `flutter packages` is used.
+- Otherwise, `flutter` is used.''';
+
+  @override
+  int get usageLineLength => _delegate.usageLineLength;
 }
 
-Future<void> pub(
-  RootConfig rootConfig,
-  String pubCommand, {
-  @required bool offline,
-  @required bool dryRun,
-  @required bool preCompile,
-}) async {
-  final pkgDirs = rootConfig.map((pc) => pc.relativePath).toList();
+/// Implementation of [Map] designed to trick [CommandRunner] into believing
+/// that any options it asks for did exist but were false.
+///
+/// This means that when [CommandRunner] checks the 'help' flag it assumes
+/// was added, it is told that the option had the value of false. This is in
+/// spite of the fact that an [allowAnything()] [ArgParser] can't have flags
+/// added and we ignored the addition of 'help' earlier in [_PubArgParser].
+/// [allowAnthing()] sets the options to be constantly empty, and for clarity
+/// we delegate to that value here, given in the constructor.
+///
+/// Conceptually, this class is a map that behaves as though it has no
+/// elements unless a specific one is requested: [containsKey] is always
+/// true, and the `[]` operator always returns a stub [_FalseOption].
+class _PubOptions implements Map<String, Option> {
```

Comment:
This part is fairly concerning, given that it doesn't (can't?) use MapMixin.

Reply:
To be frank, I wasn't aware MapMixin existed when I did this. Looking at it, I'm not sure if it would work well because it implements behaviors that'd we'd want to override and would require use to implement behavior that we don't care about. 

I think it could work but I'm not sure it would lead to a more elegant solution overall. I'd be happy to play with it, though.

Reply:
Ya - I didn't think on it to much but it did seem like it doesn't really fit the use case here. The main advantage of it is you are much much less likely to get broken by SDK rolls. Any new methods added to Map would almost certainly be implemented by the mixin in terms of the methods it asks you to implement.